servo/connectors/newrelic.py

```diff
@@ -0,0 +1,340 @@
+                raise
+
+        return [ai['host'] for ai in response.json()['application_instances']] 
+
+
+    @on_event()
+    async def measure(
+        self, *, metrics: List[str] = None, control: Control = Control()
+    ) -> Measurement:
+        """Queries Newrelic for metrics as time series values and returns a
+        Measurement object that aggregates the readings for processing by the
+        optimizer.
+
+        Args:
+            metrics (List[str], optional): A list of the metric names to measure. 
+                When None, all configured metrics are measured. Defaults to None.
+            control (Control, optional): A control descriptor that describes how            
+                the measurement is to be captured. Defaults to Control().
+
+        Returns:
+            Measurement: An object that aggregates the state of the metrics
+            queried from Newrelic.
+        """
+        if metrics:
+            metrics__ = list(filter(lambda m: m.name in metrics, self.metrics))
+        else:
+            metrics__ = self.metrics
+        measuring_names = list(map(lambda m: m.name, metrics__))
+        self.logger.info(f"Starting measurement of {len(metrics__)} metrics: {join_to_series(measuring_names)}")
+
+        start = datetime.now() + control.warmup
+        end = start + control.duration
+
+        sleep_duration = Duration(control.warmup + control.duration)
+        self.logger.info(
+            f"Waiting {sleep_duration} during metrics collection ({control.warmup} warmup + {control.duration} duration)..."
+        )
+
+        progress = DurationProgress(sleep_duration)
+        notifier = lambda p: self.logger.info(p.annotate(f"waiting {sleep_duration} during metrics collection...", False), progress=p.progress)
+        await progress.watch(notifier)
+        self.logger.info(f"Done waiting {sleep_duration} for metrics collection, resuming optimization.")
+
+        # Fetch the measurements
+        self.logger.info(f"Querying Newrelic for {len(metrics__)} metrics...")
+        readings = await self._query_newrelic(metrics__, start, end)
+        measurement = Measurement(readings=readings)
+        return measurement
+
+    async def _query_newrelic(
+        self, metrics: List[NewrelicMetric], start: datetime, end: datetime
+    ) -> List[TimeSeries]:
+        f_names_to_ms: Dict[str, List[NewrelicMetric]] = defaultdict(list)
+        fetches: Dict[str, Set[str]] = defaultdict(set)
+        for m in metrics:
+            f_names_to_ms[m.fetch_name].append(m)
+            fetches[m.fetch_name].add(m.values_selector)
+
+        newrelic_request = NewrelicRequest(base_url=self.config.api_url, fetches=fetches, start=start, end=end, step=self.config.step)
+
+        self.logger.trace(f"Getting Newrelic instance ids: {newrelic_request.url}")
+        instance_ids: List[str] = await self.dispatch_event("instances")
```

Comment:
Guessing that this probably should fall back to gathering all instances if nothing responds to the instances event to scope it? That would mean that for simpler deployments we just grab it all and you don't have to worry about instance scoping

Reply:
I'd actually been meaning to ask. Should a toggle be defined in the config as to whether the NR driver should respond the instances event? (for cases where the ec2asg connector needs to respond to it instead)

As for getting all instances, normally instances fetched from NR are filtered on a hostname which must be retrieved from OCO config (where the harness job sets it IIRC). I don't think getting all instances would provide sane behavior outside of testing environments, but am open to implementation suggestions

Reply:
See the previous reply

Reply:
discussion moved: https://github.com/opsani/servox/pull/48#discussion_r502610994